crates/stages/src/trie/mod.rs

```diff
@@ -0,0 +1,584 @@
+                    );
+
+                    let mut out = Vec::new();
+                    Encodable::encode(&value, &mut out);
+                    trie.insert(address.as_bytes().to_vec(), out)
+                        .map_err(|e| TrieError::InternalError(format!("{e:?}",)))?;
+                }
+            }
+        }
+
+        let root = H256::from_slice(trie.root().unwrap().as_slice());
+
+        Ok(root)
+    }
+
+    fn update_storage_root<DB: Database>(
+        &self,
+        tx: &Transaction<'_, DB>,
+        root: H256,
+        address: H256,
+        changed_storages: BTreeSet<H256>,
+    ) -> Result<H256, TrieError> {
+        let db = Arc::new(DupHashDatabase::new_with_root(tx, address, root)?);
+
+        let hasher = Arc::new(HasherKeccak::new());
+
+        let mut trie = PatriciaTrie::from(Arc::clone(&db), Arc::clone(&hasher), root.as_bytes())
+            .map_err(|e| TrieError::InternalError(format!("{e:?}")))?;
+        let mut storage_cursor = tx.cursor_dup_read::<tables::HashedStorage>()?;
+
+        for key in changed_storages {
+            if let Some(StorageEntry { value, .. }) =
+                storage_cursor.seek_by_key_subkey(address, key)?
+            {
+                let mut out = Vec::new();
+                Encodable::encode(&value, &mut out);
+                trie.insert(key.as_bytes().to_vec(), out)
+                    .map_err(|e| TrieError::InternalError(format!("{e:?}")))?;
+            } else {
+                trie.remove(key.as_bytes())
+                    .map_err(|e| TrieError::InternalError(format!("{e:?}")))?;
+            }
+        }
+
+        let root = H256::from_slice(trie.root().unwrap().as_slice());
+
+        Ok(root)
+    }
+
+    fn gather_changes<DB: Database>(
+        &self,
+        tx: &Transaction<'_, DB>,
+        start_tid: TransitionId,
+        end_tid: TransitionId,
+    ) -> Result<BTreeMap<H256, BTreeSet<H256>>, TrieError> {
+        let mut account_cursor = tx.cursor_read::<tables::AccountChangeSet>()?;
+
+        let mut account_changes: BTreeMap<H256, BTreeSet<H256>> = BTreeMap::new();
+
+        let mut walker = account_cursor
+            .walk(start_tid)?
+            .take_while(|res| res.as_ref().map(|(k, _)| *k < end_tid).unwrap_or_default());
+
+        while let Some((_, AccountBeforeTx { address, .. })) = walker.next().transpose()? {
+            account_changes.insert(keccak256(address), Default::default());
+        }
+
+        let mut storage_cursor = tx.cursor_dup_read::<tables::StorageChangeSet>()?;
+
+        let mut walker = storage_cursor
+            .walk((start_tid, Address::zero()).into())?
+            .take_while(|res| res.as_ref().map(|(k, _)| k.0 .0 < end_tid).unwrap_or_default());
+
+        while let Some((TransitionIdAddress((_, address)), StorageEntry { key, .. })) =
+            walker.next().transpose()?
+        {
+            account_changes.entry(keccak256(address)).or_default().insert(keccak256(key));
```

Comment:
Nit, can probably cache the `keccak`, so it is not done on every item but only when a new address comes.

Reply:
Was thinking of building the `BTreeMap` with unhashed contents, and only hash them at the end of the function. That way we only hash each address/key one time. wdyt?

Reply:
That could work well